jax/lax_linalg.py

```diff
@@ -612,13 +612,30 @@ def _lu_jvp_rule(primals, tangents):
   u_padding[-2] = (0, n - k, 0)
   u = lax.pad(jnp.triu(lu[..., :k, :]), zero, u_padding) + u_eye
 
-  la = triangular_solve(l, x, left_side=True, transpose_a=False, lower=True,
-                        unit_diagonal=True)
-  lau = triangular_solve(u, la, left_side=False, transpose_a=False,
-                         lower=False)
+  # dummy calculation to figure out which matrices u are low-rank
```

Comment:
Are you sure we need a triangular solve for this? Given that u is triangular, can’t we just check if one of the diagonal entries is zero?

Reply:
This was something we discussed when updating the gradient of `jnp.linalg.det` to work for low-rank matrices as well. There are issues with underflow so that you sometimes get infs/nans even if the diagonal isn't exactly zero. Consider this example:

```
import jax
impor jax.numpy as jnp
import struct

iota = struct.unpack('!f', bytes.fromhex('00000001'))[0]
mat = jnp.array([[iota, 0], [0, iota]])
```

If you do `iota=0` it will return False, but if you do `jax.lax_linalg.triangular_solve(mat, jnp.ones((2, 1)), left_side=True)` it will return

```
DeviceArray([[nan],
             [inf]], dtype=float32)
```

Reply:
What about using `jnp.diag(u) < jnp.ffinfo(u.dtype).eps` (or some multiple of `eps`)? I guess the `triangular_solve` is probably fine (and measures exactly what we care about).

Reply:
I'm just worried that there may be subtle edge cases where 1/eps is finite but whatever it is we're multiplying by isn't, or different backends may have subtly different behavior, and basically the safest way is to just invert it directly. I should also say that because we are only doing a triangular solve against a vector, not a matrix, it will run faster than the other triangular solves in this function.

Reply:
Great point on the vector vs. matrix valued solve!